File: AllInOne/src/Math/PowXOfN.java
/**
 * 
 * 50. Pow(x, n)
 * 
 * @author jingjiejiang
 * @history Apr 4, 2021
 * 
 */
class Solution {
    public double myPow(double x, int n) {

        assert x > -100.0 && x < 100.0;

        if (n == 0) return 1.0;
        
        double res = getHalfPow(x, n < 0 ? -n : n);

        return n < 0 ? 1 / res : res;
    }

    private double getHalfPow(double base, int factor) {

        if (factor == 0) return 1.0;

        double res = getHalfPow(base, factor / 2);

        if (factor % 2 == 0) {
            return res * res;
        } else {
            return res * res * base;
        }       
    }
}

File: AllInOne/src/Math/PowXOfN.js
/**
 *
 * 50. Pow(x, n)
 * 
 * @history Nov 25, 2022
 */
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
 var myPow = function(x, n) {
    
  if (n === 0) return 1;

  let pow = Math.abs(n);

  const res = pow % 2 === 0 ? myPow(x * x, pow / 2) : pow(x * x, (pow - 1) / 2) * x;
  
  return n < 0 ? 1 / res : res;
};